cxx-qt-lib: Implement Debug+Display for all basic types

vimpostor opened the pull request:
This completes the second item of #53.

Fixes #369

For some types I have implemented particularly fancy pretty printing, for example `QColor` **actually** prints the color via terminal escape codes, please let me know if that is "too fancy":

![Screenshot_20221221_183105](https://user-images.githubusercontent.com/21310755/209141573-41bcc48c-a6c7-43a7-838f-cbd948e89feb.png)

ahayzen commented on crates/cxx-qt-lib/src/types/qdate.rs:
```diff
@@ -37,11 +40,15 @@ mod ffi {
+
+        #[doc(hidden)]
+        #[rust_name = "qdate_to_qstring"]
+        fn qdateToQString(date: &QDate) -> QString;
```
Probably don't need a cpp bridge for this, a straight CXX binding will likely work as there is no difference between Qt 5 / 6? So something like `#[rust_name = "to_string"] fn toString(self: &QDate) -> QString` ?

vimpostor replied:
I tried that, but the Qt `QDate::toString()` doesn't have a variant with no argumens. There are variants with default arguments, but I couldn't get cxx to understand that.

The following variants are available:
- `QString toString(const QString &format, QCalendar cal = QCalendar()) const`
- `QString toString(Qt::DateFormat format = Qt::TextDate) const`
- `QString toString(QStringView format, QCalendar cal = QCalendar()) const`

I suspect we would want to use the second one,but again cxx complained that this function didn't exist if I called it with no arguments. I also tried calling it with an integer (since Qt::DateFormat is an enum) but that didn't work either, any ideas how to get it working?

ahayzen replied:
Ah right it sometimes gets confused. But see my other comment I wonder whether a generic template for all Qt types could work as well, maybe something to discuss when I'm back in Jan :-)

ahayzen commented:
Wonder if we want to use a generic template in C++ to convert T to a string, and then depend on the Qt representation of for Debug or whether we want to build them by hand for the trivial ones like QPoint etc. Something to figure out when back in Jan 🤔

ahayzen commented on crates/cxx-qt-lib/src/types/qcolor.rs:
```diff
@@ -75,6 +76,25 @@ impl Default for QColor {
+        let a = self.alpha();
+        // very simple heuristic to use a light foreground if background is dark and vice versa
+        let fg = if (r + b + g) < 384 { 255 } else { 0 };
+        // Use terminal escape codes to **actually** print the color
```
Wonder what happens on other terminals, like does this work only on a POSIX terminal or in Windows as well? Or just sh/bash/zsh? 🤔

vimpostor replied:
@ahayzen Nowadays true color support is available in pretty much every terminal emulator, see [here for a full list](https://github.com/termstandard/colors#fully-supporting).

Even the terminal application arch enemy (also known as **cmd.exe**) supports it.
The only notable terminal emulator without support is Terminal.app, but I am pretty sure that it fails silently there (i.e. the escape sequences are not shown but have no effect), so I don't find it necessar to implement true color support detection even though that would be very easy to do (see link above).

ahayzen replied:
There is Debug Vs Display ... "Debug should format the output in a programmer-facing, debugging context" sounds like it could be with colour in a terminal 🤔 Maybe we should see what other crates do  ...

vimpostor replied:
I think that sounds like a good compromise.
I have updated it to pretty-print with colors only in the Debug formatter and the Display formatter prints it without colors now.

Theoretically it would be also possible to detect if the output is a terminal (see `isatty(3)`), but let's not overengineer this and it would work only on Unix anyway.

ahayzen-kdab commented on crates/cxx-qt-lib/src/types/qrect.rs:
```diff
@@ -73,6 +74,19 @@ impl Default for QRect {
 }
 
+impl fmt::Display for QRect {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
```
should we use Qt for the display format ? As most Qt types have `QDataStream &operator<<(QDataStream &stream, const T &point)`  eg i wonder if this would work ... 

Add the following to  `crates/cxx-qt-lib-headers/include/common.h` ...
```cpp
template<typename T>
QString
display(const T& value)
{
  return QString() << value;
}
```

Then add that to the bridge
```rust
        #[doc(hidden)]
        #[rust_name = "qrect_display"]
        fn display(value: &QRect) -> QString;
```

Then in the `fmt` methods do
```rust
write!(f, "{}", qrect_display(self))
```

vimpostor replied:
That's a good idea.
The simple `QString() << value;` that you proposed doesn't quite work, but I was able to get it working by constructing a `QDebug` object that writes into a `QString` and streaming (`<<`) into that instead.

ahayzen-kdab commented:
Also note there are more types added in `main` when rebasing :-) (sorry)

vimpostor commented:
> Also note there are more types added in `main` when rebasing :-) (sorry)

No worries, I implemented `Display` for the new types `{QByteArray, QStringList, QVector[2-4]D}` and fixed the merge conflicts too.

ahayzen-kdab commented:
So the other option for `QByteArray` could be to keep the Qt representation for Display/Debug.

But then have something like
```rust
impl TryFrom<&QByteArray, Utf8Error> for String {
...  // try to convert to UTF8 String fail otherwise
}

impl QByteArray {
   pub fn to_utf8(&self) -> Result<String, Utf8Error> {

   }
}
````

Although `str` also already has `str::from_utf8` which can take a `&[u8]` which we can already reach via a slice of a QByteArray :thinking:  So we could change the test to use `str::from_utf8(bytearray.as_slice()).unwrap()` ? :thinking: 

https://doc.rust-lang.org/stable/std/str/fn.from_utf8.html

Or do we feel that having the `to_string() -> String` not be the Qt representation and be a UTF8 string more natural ?

vimpostor commented on crates/cxx-qt-lib-headers/include/common.h:
```diff
@@ -7,6 +7,9 @@
 
 #pragma once
 
+#include <QDebug>
```
Done

vimpostor commented:
> (and there is `String::from_utf8` from a `Vec<u8>`

That's what is already being used since 0a5684b0, I didn't feel it necessary to replace it with the Qt formatter.
If you want to favor consistency however, I can surely replace that with the Qt formater if you want.

ahayzen-kdab reviewed (approved):
Lets land this and then we can modify any specific ones further if we need to :-)
